OrbitCore/SymbolHelper.cpp

```diff
@@ -41,118 +44,117 @@ std::vector<std::string> ReadSymbolsFile() {
   }
 
-  std::vector<std::string> directories;
+  std::vector<path> directories;
   std::fstream infile(file_name);
   if (!infile.fail()) {
     std::string line;
     while (std::getline(infile, line)) {
       if (absl::StartsWith(line, "//") || line.empty()) continue;
 
-      const std::string& dir = line;
-      if (Path::DirExists(dir)) {
+      const path& dir = line;
+      if (std::filesystem::exists(dir)) {
         directories.push_back(dir);
       } else {
-        ERROR("Symbols directory \"%s\" doesn't exist", dir);
+        ERROR("Symbols directory \"%s\" doesn't exist", dir.string());
       }
     }
   }
   return directories;
 }
 
-ErrorMessageOr<std::string> FindSymbolsFile(const std::string& module_path,
-                                            const std::vector<std::string>& search_paths,
-                                            const std::string& build_id) {
-  std::string filename = Path::GetFileName(module_path);
-  std::string filename_without_extension = Path::StripExtension(filename);
+ErrorMessageOr<void> VerifySymbolsFile(const path& symbols_path, const std::string& build_id) {
+  OUTCOME_TRY(symbols_file, ElfFile::Create(symbols_path.string()));
 
-  std::vector<std::string> search_file_paths;
-  search_file_paths.push_back(module_path);
+  if (!symbols_file->HasSymtab()) {
+    return ErrorMessage(absl::StrFormat("Elf file \"%s\" does not contain symbols.", symbols_path.string()));
+  }
+  if (symbols_file->GetBuildId().empty()) {
+    return ErrorMessage(
+        absl::StrFormat("Symbols file \"%s\" does not have a build id", symbols_path.string()));
+  }
+  const std::string& file_build_id = symbols_file->GetBuildId();
+  if (build_id != file_build_id) {
+    return ErrorMessage(
+        absl::StrFormat(R"(Symbols file "%s" has a different build id: "%s" != "%s")", symbols_path.string(),
+                        build_id, file_build_id));
+  }
+  return outcome::success();
+}
 
-  for (const auto& directory : search_paths) {
-    search_file_paths.emplace_back(
-        Path::JoinPath({directory, filename_without_extension + ".debug"}));
-    search_file_paths.emplace_back(Path::JoinPath({directory, filename + ".debug"}));
-    search_file_paths.emplace_back(Path::JoinPath({directory, filename}));
+ErrorMessageOr<path> FindSymbolsFile(const path& module_path,
+                                     const std::vector<path>& search_directories,
+                                     const std::string& build_id) {
+  if (build_id.empty()) {
+    return ErrorMessage(absl::StrFormat(
+        "Could not find symbols file for module \"%s\", because it does not contain a build id",
+        module_path.string()));
   }
 
-  LOG("Trying to find symbols for module: \"%s\"", module_path);
-  for (const auto& symbols_file_path : search_file_paths) {
-    if (!std::filesystem::exists(symbols_file_path)) continue;
+  const path& filename = module_path.filename();
+  path filename_dot_debug = filename;
```

Comment:
nit, the variables could be const, right?

Reply:
no unfortunately they can't. the `replace_extension()` method is not const. I did not find a better way to add/replace the extension in the world of std::filesystem.

Reply:
okay, thanks for the explanation.